components/script/dom/bindings/utils.rs

```diff
@@ -42,34 +43,29 @@ use crate::dom::bindings::error::throw_invalid_this;
 use crate::dom::bindings::inheritance::TopTypeId;
 use crate::dom::bindings::str::DOMString;
 use crate::dom::bindings::trace::trace_object;
-use crate::dom::windowproxy;
+use crate::dom::windowproxy::{WindowProxyHandler, PROXY_HANDLER};
 use crate::script_runtime::JSContext as SafeJSContext;
 
-/// Proxy handler for a WindowProxy.
-pub struct WindowProxyHandler(pub *const libc::c_void);
-
-impl MallocSizeOf for WindowProxyHandler {
-    fn size_of(&self, _ops: &mut MallocSizeOfOps) -> usize {
-        // FIXME(#6907) this is a pointer to memory allocated by `new` in NewProxyHandler in rust-mozjs.
-        0
-    }
-}
-
-#[derive(JSTraceable, MallocSizeOf)]
+#[derive(JSTraceable)]
 /// Static data associated with a global object.
 pub struct GlobalStaticData {
     /// The WindowProxy proxy handler for this global.
-    pub windowproxy_handler: WindowProxyHandler,
+    pub windowproxy_handler: Arc<WindowProxyHandler>,
 }
 
 impl GlobalStaticData {
     /// Creates a new GlobalStaticData.
     pub fn new() -> GlobalStaticData {
         GlobalStaticData {
-            windowproxy_handler: windowproxy::new_window_proxy_handler(),
+            windowproxy_handler: Arc::new(WindowProxyHandler::new(&PROXY_HANDLER)),
         }
     }
 }
+impl MallocSizeOf for GlobalStaticData {
```

Comment:
This looks the same as a derived implementation. Why can't we use the derived version instead?

Reply:
We tried that first but got an error with the derive macro. According to the documentation[^1] it seems like it is intentional that this case must be decided manually. We just passed it through without thinking about it too much as the impl for `ProxyWindowHandler` returns `0` at the moment anyway and this copy of the `Arc` is the only one that would be counted because the other copies (that only exist) in `WindowProxy` are skipped in the derive macro through our annotation.

However in some cases (e.g. `new_dissimilar_origin` constructor) there is no window so the `Arc` would not be counted at all (right now it would return 0 anyway as mentioned before but this might lead to issues when someone eventually fixes #6097). Do you know a better solution? Anyway, thanks for pointing it out. I actually just noticed that I **missed a leak [in the `set_window` method](https://github.com/servo/servo/blob/6b8804d263ac3f2fa5835116f947fc79f7bd1ab4/components/script/dom/windowproxy.rs#L636)**. Here the window that the `WindowProxy` refers to is swapped and entirely new `WindowProxyHandler` is created. Now I'm wondering whether `Arc` was necessary in the first place or if we should just make deep copies instead...? 🤔

[^1]: https://doc.servo.org/malloc_size_of/index.html

Reply:
Ah, I see. The usual path here is to use a derive and add an `#[ignore_malloc_size_of]` annotation on the field that causes the error. This means that any new fields added to the struct in the future won't be forgotten.

Reply:
Ok, changed that. I also tried to fix the leak in `set_window` (see 2nd commit) but when I switch websites a bit now, I get a segfault. Will have to check that out in the coming week. If you want to merge the PR before that, we can drop the second commit for now and keep the third leak for now.

Reply:
I added a few log messages about when `set_window` is called and when instances of `WindowProxyHandler` are created and destroyed. Clicking around a bit until SEGFAULT (<http://example.com> -> <http://blog.fefe.de> -> a few subpages there, maybe also back/forwards arrow) got me this log:

```
Acquired window proxy handler 0x512000206a40
Acquired window proxy handler 0x512000294940
Setting window -> potential drop of window proxy Handler
Acquired window proxy handler 0x5120005a67c0
Acquired window proxy handler 0x5120004ec340
Setting window -> potential drop of window proxy Handler
Acquired window proxy handler 0x5120004eba40
Acquired window proxy handler 0x5120003a72c0
Setting window -> potential drop of window proxy Handler
Acquired window proxy handler 0x51200028da40
Dropping window proxy handler 0x5120004eba40
Caught signal 11 in thread "Script(1,2)"
Servo was terminated by signal 11
```

Reply:
Thinking about it now, maybe it's better to hand the `WindowProxyHandler` pointer to the GC. My reasoning here is that the GC'ed `WindowProxy` (in the `NewWindowProxy` C++ method) stores a pointer to the handler and I'm not sure if this stored pointer can escape in the C++ code and hence live longer than `Window` or `WindowProxy`. What do you think?